### ISSUE TITLE
MaxLineLength on test function names

### ISSUE THREAD

**Issue by tasomaniac** [detekt/detekt#2745]
## Expected Behavior of the rule

MaxLineLength rule already defines some places where it can be ignored 

```
    excludePackageStatements: true
    excludeImportStatements: true
    excludeCommentStatements: false
```

Another one we are seeing is the test function names. In Kotlin, we are using back-quotues (`) and test names are English sentences. That's why they tend to go long but that is not a bad thing. 

We also don't want to disable this check in all test code since it is a good one.

Would be good to somehow exclude those test names from this check.

**Comment by cortinico** [comment]
There are several options to tackle this:

1. Add a `excludeBackticksStatements` config property. That will exclude lines that contains 2 backticks and will cover this scenario (but will open to other false negatives that we probably want to catch).

2. Add a `excludeBackticksFunctionName` config property. That will check if `line.contains("fun `")`. This will probably work but it's not really elegant plus it will probably create a precedent for adding several edge-case config property on this rule.

3. You could exclude this rule on your tests:
https://github.com/detekt/detekt/blob/664e734e4ddb04e9f578cac3bc9d22cc118d56c7/config/detekt/detekt.yml#L116-L117
This will unfortunately disable the rule at all on all your tests.

Happy to hear what's your/other's opinion on this.

**Comment by tasomaniac** [comment]
Excluding in all test code is possible at the moment. It is ok but not the best solution. 

It's a common practice to use back ticks for test names but they can potentially also be in the production code. Not possible in Android but it is on jvm. 

One could argue that it is also fine to exclude them in production code since these are **impossible** to make it shorter if their name makes sense and they're just long because of their nature.

**Comment by schalkms** [comment]
Thanks @cortinico for describing the solution concept for this issue. 
@tasomaniac Is there anything that should be changed? As I understand, the existing detekt features cover the use cases, as described above by Nico. Can we close this then?

**Comment by tasomaniac** [comment]
Are you sure? He was talking about potential solutions. I don't think those options are available atm

**Comment by tasomaniac** [comment]
That's not a solution since it disables the check for the whole test sources. It's more like a workaround. 

Imagine you have a test file with 200 lines and 1 test name is longer. Just to get rid of 1 warning, you will have to ignore this check for all 200 lines.

**Comment by schalkms** [comment]
Then options 3 is canceled out. 
Could we exclude ‘’ altogether from this rule?

**Comment by maltaisn** [comment]
> In Kotlin, we are using back-quotues (`) and test names are English sentences. That's why they tend to go long but that is not a bad thing.

> One could argue that it is also fine to exclude them in production code since these are **impossible** to make it shorter if their name makes sense and they're just long because of their nature.

If implemented, this shouldn't only apply to function names with backticks. One doesn't have to use backticks to have a long function name that's impossible to shorten. Android developers will know, tests run on an Android device (usually in the `androidTest` source set) cannot have some of the special characters allowed with backticks, so normal function names must be used.

```
// A bit far-fetched--but not entirely impossible--function in androidTest source set.
fun should_run_room_migration_dropping_old_tables_adding_new_columns_to_existing_table_and_setting_default_values_on_employee_entities()
```

Instead, I would opt for a combination of the existing `FunctionMaxLength` and a new `MaxLineLength.excludeFunctionDeclarations`. The former can already be restricted to specific source sets so one could rely on `FunctionMaxLength` for catching long lines with function declarations in the main source set while allowing any length in test source sets. The only shortcoming to this approach is that if someone decides to declare a function starting at column 100 and exceeds the line length, no warning will be shown.

```yaml
style:
  MaxLineLength:
    active: true
    maxLineLength: 120
    excludeFunctionDeclarations: true
naming:
  FunctionMaxLength:
    active: true
    excludes: ['**/test/**', '**/androidTest/**', '**/jvmTest/**', '**/jsTest/**', '**/iosTest/**']
    maximumFunctionNameLength: 30
```
I haven't checked if that's possible to implement though. What do you think?

**Comment by cortinico** [comment]
> What do you think?

This is really similar to option 2 of my suggestion `excludeBackticksFunctionName`. However, you suggest to exclude all the function names.

I'd still prefer an option to exclude only function declarations that has backticks. The reason behind my preference is that function names with backticks are longer by their nature (you write natural language inside with several spaces). The function name your presented as an argument is possible, but should be discouraged (hence flagged by detekt). Specifically, I would argue that in that example you should replace snake case with camel case and shorten the function name.

**Comment by BraisGabin** [comment]
If we allow to define a function with a name of 140 chars we should allow to have lines with more than 120 chars when you are calling it. We could say that if you want to use it you should use an import alias, but that sounds odd.

If we don't think about how to call that function we are just thinking about a problem that appears when we use JUnit. And I think that it's better to keep the standard rules as much library-independent as possible.

And only allowing backticks is not enought because of `androidTest`, as @maltaisn's comment points out.

**Comment by tasomaniac** [comment]
Good point. Test functions are never called explicitly. That's why long names are just fine.

I don't agree that this is about making it library dependant. This is not only about jUnit. The request is generally true with all test frameworks.

**Comment by BraisGabin** [comment]
Now with #4102 you can add to `MaxLineLength` configuration `ignoreAnnotated: ['Test']`. It is not the exact requested behaviour because it will ignore all the code inside the function but I think that it's better than exclude all the code in the test folder.

**Comment by tasomaniac** [comment]
That's awesome. Thanks. This would satisfy the needs of the original issue. I would close it but I will leave that up to you.
